style: Flow rewrite to avoid skip of init or unreachable code warnings

LoveKarlsson opened the pull request:
Rewrote code with gotos or if(true) that caused warnings of 
unreachable code or bypassing of variable initialization.

Example of IAR warnings:
                goto bail;
                ^
"~\zephyrproject\zephyr\drivers\timer\nrf_rtc_timer.c",615  Warning[Pe546]: 
   transfer of control bypasses initialization of:
            variable "now" (declared at line 626)

LoveKarlsson committed:
style: Flow rewrite to avoid skip of init or unreachable code warnings

Rewrote code with gotos or if(true) that caused warnings of
unreachable code or bypassing of variable initialization.

Signed-off-by: Lars-Ove Karlsson <lars-ove.karlsson@iar.com>

jakub-uC commented:
The changes look good from a shell perspective. However, I'd prefer to have the kernel owner review and approve it before moving forward. Appreciate your understanding!

LoveKarlsson commented:
> The changes look good from a shell perspective. However, I'd prefer to have the kernel owner review and approve it before moving forward. Appreciate your understanding!

Absolutely! We might add a comment to explain that it is bailing for example, just to be clear with what is happening.


I'm still struggling anyway with fixing the commit message (there is always something wrong with it, but I'm new to this :) ) so I'm in no hurry

LoveKarlsson committed:
style: Flow rewrite to avoid skip of init or unreachable code warnings

Rewrote code with gotos or if(true) that caused warnings of
unreachable code or bypassing of variable initialization.

Signed-off-by: Lars-Ove Karlsson <lars-ove.karlsson@iar.com>

peter-mitsis commented:
Functionally, everything looks good. However, I do have a couple of nitpicky items.

1. There appears to be a commit that contains no changes. I assume that this is a remnant of a rebase of some kind--it should be removed.
2. Although the commit message references "if (true)" and "unreachable code", I do not observe either of those cases in the commit. Let's clean that up.
3. Instead of adding the indentation from "else if" and "else", can we not just separate the variable declaration and assignment so that the declaration is done at a common location before the "goto" or "return" and the location of the assignment remains unchanged? (I ask because it could help keep indentation down, reduce the number of lines to inspect and it could offer some insight into what the IAR tool catches so that offending patterns can be avoided when writing  future code)

LoveKarlsson commented:
> Functionally, everything looks good. However, I do have a couple of nitpicky items.
> 
> 1. There appears to be a commit that contains no changes. I assume that this is a remnant of a rebase of some kind--it should be removed.
> 2. Although the commit message references "if (true)" and "unreachable code", I do not observe either of those cases in the commit. Let's clean that up.
> 3. Instead of adding the indentation from "else if" and "else", can we not just separate the variable declaration and assignment so that the declaration is done at a common location before the "goto" or "return" and the location of the assignment remains unchanged? (I ask because it could help keep indentation down, reduce the number of lines to inspect and it could offer some insight into what the IAR tool catches so that offending patterns can be avoided when writing  future code)

1. That  was my fault while trying to change the log message to conform. GIT and I are not the best of friends

2. The if (true) was to be short for the if (CONFIG(...)) that resulted in a constant value.

3. That is a good approach. Moving the declarations to the start of the function would avoid this warning. 

P.S.
I still don't like gotos but that is just my opinion :)

cfriedt commented on subsys/shell/shell.c:
```diff
@@ -1414,15 +1414,14 @@ void shell_uninit(const struct shell *sh, shell_uninit_cb_t cb)
-	if (cb) {
-		cb(sh, err);
 	} else {
-		__ASSERT_NO_MSG(0);
+		int err = instance_uninit(sh);
+
+		if (cb) {
+			cb(sh, err);
+		} else {
+			__ASSERT_NO_MSG(0);
+		}
```
Is this really necessary? Returning early from a function shouldn't require an additional level of nesting.

If the issue is that IAR doesn't allow local variables to be declared after control statements, then couldn't the declaration / initialization of the variable just be moved to the top of the function?

Alternatively, is there some flag that can be passed to IAR to allow for C99-style variable declaration? It would be a shame to have to regress to C89 to satisfy one compiler.

LoveKarlsson replied:
It is only a warning as IAR defaults to C99. It only warns that the initialization might not occur due to an earlier return or goto, especially if it is dynamically initialized with a return value from a call with potential side effects.

We already have disabled quite a lot of warnings and errors to be able to compile zephyr. This one was just one that I thought wouldn't have to occur, but sure, it can be disabled.

LoveKarlsson commented:
I think I'm going to close this pull request. 

Would it be of interest to get a summary of code constructs the IAR compilers gives warnings or errors for, and if so, where to post and discuss them?

LoveKarlsson commented:
Closing
